File: core/src/com/heroes3/livewallpaper/AssetsParser/PngWriter.java
package com.heroes3.livewallpaper.AssetsParser;

import java.io.*;
import java.nio.ByteBuffer;
import java.util.Arrays;
import java.util.zip.CRC32;
import java.util.zip.Deflater;
import java.util.zip.DeflaterInputStream;

class PngWriter {
    private ByteArrayOutputStream out = new ByteArrayOutputStream();

    byte[] create(int width, int height, byte[] palette, byte[] transparent, byte[] data) throws IOException {
        out.write(new byte[]{-119, 80, 78, 71, 13, 10, 26, 10});
        writeChunk("IHDR", createIHDR(width, height));
        writeChunk("PLTE", palette);
        writeChunk("tRNS", transparent);
        writeChunk("IDAT", createIDAT(data, width, height));
        writeChunk("IEND", new byte[0]);
        return out.toByteArray();
    }

    private byte[] createIDAT(byte[] data, int width, int height) throws IOException {
        ByteArrayOutputStream scanlines = new ByteArrayOutputStream();
        for (int i = 0; i < height; i++) {
            scanlines.write((byte) 0);
            scanlines.write(
                Arrays.copyOfRange(data, width * i, width * i + width)
            );
        }

        byte[] input = scanlines.toByteArray();
        byte[] output = new byte[input.length];
        Deflater deflater = new Deflater();
        deflater.setInput(input);
        deflater.finish();
        deflater.deflate(output);
        deflater.end();
        return output;
    }

    private byte[] createIHDR(int width, int height) throws IOException {
        ByteArrayOutputStream ihdr = new ByteArrayOutputStream();
        ihdr.write(
            ByteBuffer
                .allocate(4)
                .putInt(width)
                .array()
        );
        ihdr.write(
            ByteBuffer
                .allocate(4)
                .putInt(height)
                .array()
        );
        ihdr.write((byte) 8); // bitDepth
        ihdr.write((byte) 3); // colorType
        ihdr.write(0); // compressionMethod
        ihdr.write(0); // filterMethod
        ihdr.write(0); // interlaceMethod
        return ihdr.toByteArray();
    }

    private void writeChunk(String type, byte[] content) throws IOException {
        out.write(
            ByteBuffer
                .allocate(4)
                .putInt(content.length)
                .array());
        out.write(type.getBytes());
        out.write(content);
        out.write(
            ByteBuffer
                .allocate(4)
                .putInt((int) getCRC(type, content))
                .array()
        );
    }

    private long getCRC(String type, byte[] content) {
        CRC32 crc32 = new CRC32();
        crc32.update(type.getBytes(), 0, type.getBytes().length);
        crc32.update(content, 0, content.length);
        return crc32.getValue();
    }
}

File: core/src/com/heroes3/livewallpaper/parser/formats/PngWriter.kt
package com.heroes3.livewallpaper.parser.formats

import java.io.ByteArrayOutputStream
import java.io.IOException
import java.nio.ByteBuffer
import java.util.zip.CRC32
import java.util.zip.Deflater

internal class PngWriter {
    private val output = ByteArrayOutputStream()

    @Throws(IOException::class)
    fun create(width: Int, height: Int, palette: ByteArray, transparent: ByteArray, data: ByteArray): ByteArray {
        output.write(byteArrayOf(-119, 80, 78, 71, 13, 10, 26, 10))
        writeChunk("IHDR", createIHDR(width, height))
        writeChunk("PLTE", palette)
        writeChunk("tRNS", transparent)
        writeChunk("IDAT", createIDAT(data, width, height))
        writeChunk("IEND", ByteArray(0))
        return output.toByteArray()
    }

    private fun createIDAT(data: ByteArray, width: Int, height: Int): ByteArray {
        val scanlines = ByteArrayOutputStream()
        for (i in 0 until height) {
            scanlines.write(0)
            scanlines.write(
                data.copyOfRange(width * i, width * i + width)
            )
        }
        val input = scanlines.toByteArray()
        val output = ByteArray(input.size + 100) // TODO fix hardcoded value
        val deflater = Deflater()
        deflater.setInput(input)
        deflater.finish()
        deflater.deflate(output)
        deflater.end()
        return output
    }

    private fun createIHDR(width: Int, height: Int): ByteArray {
        val ihdr = ByteArrayOutputStream()
        ihdr.write(
            ByteBuffer
                .allocate(4)
                .putInt(width)
                .array()
        )
        ihdr.write(
            ByteBuffer
                .allocate(4)
                .putInt(height)
                .array()
        )
        ihdr.write(8) // bitDepth
        ihdr.write(3) // colorType
        ihdr.write(0) // compressionMethod
        ihdr.write(0) // filterMethod
        ihdr.write(0) // interlaceMethod
        return ihdr.toByteArray()
    }

    private fun writeChunk(type: String, content: ByteArray) {
        output.write(
            ByteBuffer
                .allocate(4)
                .putInt(content.size)
                .array())
        output.write(type.toByteArray())
        output.write(content)
        output.write(
            ByteBuffer
                .allocate(4)
                .putInt(getCRC(type, content).toInt())
                .array()
        )
    }

    private fun getCRC(type: String, content: ByteArray): Long {
        val crc32 = CRC32()
        crc32.update(type.toByteArray(), 0, type.toByteArray().size)
        crc32.update(content, 0, content.size)
        return crc32.value
    }
}